src/JsonApi/Serializer/ItemNormalizer.php

```diff
@@ -92,6 +92,19 @@ public function normalize($object, $format = null, array $context = [])
         $allRelationshipsData = $this->getComponents($object, $format, $context)['relationships'];
         $populatedRelationContext = $context;
         $relationshipsData = $this->getPopulatedRelations($object, $format, $populatedRelationContext, $allRelationshipsData);
+
+        if (!isset($context['api_included_resources'])) {
+            // initialize tracking of already included resources
+            $context['api_included_resources'] = [];
+        }
+
+        // do not include primary resources
+        $context['api_included_resources'] = [
+            $resourceMetadata->getShortName() => [
```

Comment:
What about self-referencing relationships?

Reply:
Sorry, can't follow you. Maybe the comment above i misleading. It's just about primary data is already part of the "regular" answer and it's not necessary to put them into the `included` part too.

Reply:
@dunglas Do you want me to do or change anything?